### DISCUSSION TITLE
Calculation using multi-select

### DISCUSSION THREAD

**Discussion by PaulMcF1987** [Budibase/budibase#6827] [Help]
I am making a calculator so that customers can go on and calculate a rough estimate for their job.

I have added an `additional options` section using the multi-select and I have set it up as check boxes rather than the dropdown.

My question is...

what is the best way to calculate a cost based on the checkboxes that are selected... For example...

    Box A - £20
    Box B - £25
    Box C - £30
    Box B - £35

If customer selects `box A` Add £20... If the select `Box A and Box C` add £20 + £30 = £50

I was thinking something along the lines of 

        let price = 0;
            if ($("Form.Fields.FieldValue") == "Box A") {
            price = 20
            } 
            else if ($("Form.Fields.FieldValue") == "Box A" && $("Form.Fields.FieldValue") == "Box C) {
            price = 20 + 30
            } 
            return price;

However, that seems quite messy and Iit doesnt work, there is no JS Error which suggests that the function is fine.

The problem I think is that it might be stored as `"0":"Box A","1":"Box C"` ... and the problem I would have with this is that the number before the `Box X` will not always be the same depending on the order which the boxes are ticked.

Has anyone come up against this and managed to create a solution?

**Comment by melohagan** [comment]
Hey @PaulMcF1987 

I've added a tutorial to the docs: https://docs.budibase.com/docs/multi-select-picker#tutorial-sum-of-selected-items

Hopefully that gives you some idea of how to accomplish this!

**Reply by PaulMcF1987** [reply]
@melohagan Ive had a look at the solution again. It might work to a degree, however, the values associated with the checkboxes are variable depending on the other information so Ideally Id need something a bit more flexible.

For example if `x=y then checkboxA = C else If X=z then CheckboxA = D`

And from there I can use the value for CheckboxA to complete the rest of the calculation.

Is there a viable solution for this?
I have tried multiple different things using `primaryDisplay` etc but cannot get a solution that populates an answer.

**Comment by PaulMcF1987** [comment]
@melohagan

Thank you for your reply, although, Im not 100% sure this would work for me.

the multi select is only a small part of the whole calculator...

I have created a blank table (not intending on saving any values to it, table is purely to allow the widgets to be used...

I have set one of the fields as a `multi-select` and populated it with the options.

I was then looking to add some JS which, onchange, then populates a different column with values that have been stipulated in the JS as mentioned above in the original question.

There are a few other dropdowns (`option-pickers`) which populate other boxes.

I then have a calculate button which will take the numbers that have been populated and add them together to give a total

![image](https://user-images.githubusercontent.com/75368133/180032094-2323a1da-44c2-4cc2-a386-4ee8bf6d56fa.png)

The options in black determine what value goes into the top box
The items in red derermine what value goes into the middle box
The checkbox at the end then determines what value goes into the bottom box.

The calculate button then adds them together.

Is there a way to differentiate between the selected boxes so that I can write a JS formula? Or do something similar to the link you posted that will allow for the values calculated to be then used for a further calculation?

**Reply by melohagan** [reply]
Hey @PaulMcF1987 have you considered using <a href="https://docs.budibase.com/docs/app-state">App State</a> to keep track of the various option picker states.

You can create a map in JavaScript. So your **On Change** events could add to their respective entries in the App State.
Something like this binding for the **Update State** action:
```
let pickerState = $("State.Pickers);
if (pickerState) {
  pickerState["additionalOptions"] = $("Form.Fields.FieldValue")
}
return pickerState;
```
You can choose whether or not to persist state, i.e. if you navigate away from the page will it remember your choice, or will it be refreshed each time you visit that screen.

Then your associated text box would have the binding `{{ State.Pickers.additionalOptions }}`
Anytime the app state is updated, then the value in your text boxes would also be updated.

**Reply by PaulMcF1987** [reply]
Thanks @melohagan 
How would I set the values using this though? With the field value being a name I need to be able to attribute a value to each of the named options dependant on the options which were selected before.

For example...
        
    I was a gardener the first dropdown might be garden size... Small or large.
        
    The select boxes might then read `Grass cutting`, `weeding`, `hedge trimming`
    
The select boxes would then be priced differently depending on whether the garden is small or large.
And which ever options are chosen would be added together to give a total.

Would that be achievable using the example that you mentioned above?
If so, how do I reference the select box options in the formula?

**Reply by PaulMcF1987** [reply]
@melohagan I have been looking at this and Ill be honest and say, I havent got a clue where to start.
the main reason being, I dont know how to reference the checkboxes in the formula.

But that being said, I feel as though this is getting a lot more complicated than it needs to be. If there was a way for me to reference the checkboxes in JS then I could add them to the initial formula as intended.

**Reply by PaulMcF1987** [reply]
@melohagen

I have tried replacing the multi-select with separate boolean fields.
I have then tried using `if form.fields.x == "1" { x = y}....` but it doesnt seem to work. How are the boolean values saved in the table? I have checked the table in phpmyadmin and the output seems to be saved as a `1`
Should I be using a different argument?

I have also tried, instead of 1s and 0s using `checked and unchecked` and `true and false` but none of these work

also tried `form.fields.x.checked`
